### PR TITLE
[SYCL] Key/Value sorting with fixed-size private array input

### PR DESCRIPTION
**Pull request by againull**
Implementation of sort_key_value_over_group APIs from 
https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc#functions-with-fixed-size-arrays

### CODE REVIEW

**Comment by againull** [comment]
Tag @andreyfe1

**Review by andreyfe1 — COMMENTED** [review]
Looks good to me. Thanks!

**Comment by andreyfe1** [comment]
Regarding the comment here: https://github.com/intel/llvm/pull/13942#discussion_r1634504220
Does oneDPL `zip_iterator`'s support plan to be addressed in this or any other PR?

**Comment by againull** [comment]
> Regarding the comment here: [#13942 (comment)](https://github.com/intel/llvm/pull/13942#discussion_r1634504220) Does oneDPL `zip_iterator`'s support plan to be addressed in this or any other PR?

This PR adds support for zip iterator. Unfortunately I can't add a test  for that to this repo because it will require creating a dependency from OneDPL.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 1]
```diff
@@ -50,8 +51,117 @@ static __SYCL_ALWAYS_INLINE T *align_scratch(sycl::span<std::byte> scratch,
...
+  }
+  // Broadcast leader's pointer (the beginning of the scratch) to all work
+  // items in the group.
+  keys_scratch_begin = sycl::group_broadcast(g, keys_scratch_begin);
+  values_scratch_begin = sycl::group_broadcast(g, values_scratch_begin);
+  return std::make_pair(keys_scratch_begin, values_scratch_begin);
+}
 #endif
 
+template <typename T1, typename T2> class key_value_iterator {
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/include/sycl/detail/group_sort_impl.hpp]: Ok, moved to a separate file.

File: sycl/include/sycl/ext/oneapi/experimental/group_sort.hpp [hunk 2]
```diff
@@ -239,6 +254,63 @@ sort_key_value_over_group(
...
+std::enable_if_t<
+    sycl::ext::oneapi::experimental::is_property_list_v<
+        std::decay_t<Properties>> &&
+        !sycl::ext::oneapi::experimental::is_property_list_v<
+            std::decay_t<Compare>> &&
+        !detail::is_array_key_value_sorter<Compare, Group, T, U, Properties,
+                                           ElementsPerWorkItem>::value,
+    void>
+sort_key_value_over_group(
+    experimental::group_with_scratchpad<Group, Extent> exec,
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/include/sycl/ext/oneapi/experimental/group_sort.hpp]: Changed to support all group helpers.

File: sycl/include/sycl/ext/oneapi/experimental/group_sort.hpp [hunk 3]
```diff
@@ -239,6 +254,63 @@ sort_key_value_over_group(
...
+          typename Properties = ext::oneapi::experimental::empty_properties_t>
+std::enable_if_t<
+    sycl::ext::oneapi::experimental::is_property_list_v<
+        std::decay_t<Properties>> &&
+        detail::is_array_key_value_sorter<ArraySorter, Group, T, U, Properties,
+                                          ElementsPerWorkItem>::value,
+    void>
+sort_key_value_over_group(Group group, sycl::span<T, ElementsPerWorkItem> keys,
+                          sycl::span<U, ElementsPerWorkItem> values,
+                          ArraySorter array_sorter,
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/include/sycl/ext/oneapi/experimental/group_sort.hpp]: Ok, renamed to Sorter.

File: sycl/include/sycl/detail/key_value_iterator.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  std::tuple<T1 *, T2 *> KeyValue;
+};
+
+template <typename T> void swap(T &first, T &second) {
+  std::swap(first, second);
+}
+
+} // namespace detail
+} // namespace _V1
+} // namespace sycl
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/detail/key_value_iterator.hpp:94]: nit: add empty line at the EOF

**Review by dm-vodopyanov — APPROVED** [review]

File: sycl/include/sycl/detail/key_value_iterator.hpp (hunk 4 above)
Comment (againull) [reply sycl/include/sycl/detail/key_value_iterator.hpp:94]: Thank you, will have to move extension doc from proposed to supported in another PR, will make this correction there.

**Pull request merged**